### ISSUE TITLE
`INSERT` statements with many parameters: `'h' format requires -32768 <= number <= 32767`

### ISSUE THREAD

**Issue by dataders** [aws/amazon-redshift-python-driver#165]
executing prepared `INSERT` statements with large numbers of parameters will fail with the following error message

the error message is
> `'h' format requires -32768 <= number <= 32767`

this error is encountered for code that previously worked without issue with `psycopg2` as the driver.

@elongl @jaklan and @misteliy surfaced this issue via https://github.com/dbt-labs/dbt-redshift/issues/347 

this repo's "cousin" has some more context in this particular issue comments (below). My (limited) understanding is that there's a less fancy type inference happening related to the types of parameter sets passed to `.execute()`

- https://github.com/tlocke/pg8000/issues/120#issuecomment-1288381859
- https://github.com/tlocke/pg8000/issues/120#issuecomment-1293946586

### possible solution
implement `Cursor.setinputsizes()`, which is currently stubbed out

https://github.com/aws/amazon-redshift-python-driver/blob/910be71314229e5b2febc60ddf8b8bcc992ea5f0/redshift_connector/cursor.py#L471-L476

perhaps `pg8000`'s implementation could be useful here?
https://github.com/tlocke/pg8000/blob/f751b94a910975fd2e551814b7395336a5153cc8/pg8000/dbapi.py#L616-L629

```
    def setinputsizes(self, *sizes):
        """This method is part of the `DBAPI 2.0 specification"""
        oids = []
        for size in sizes:
            if isinstance(size, int):
                oid = size
            else:
                try:
                    oid = PY_PG[size]
                except KeyError:
                    oid = UNKNOWN
            oids.append(oid)

        self._input_oids = oids
```

**Comment by Brooke-white** [comment]
Hey @dataders , thank you for opening this issue. The team will take a look into this and work on reproducing as well as evaluating how pg8000 addressed this fix.

**Comment by jiezhen-chen** [comment]
@dataders Thanks for writing up this issue! Could you provide a reprex for this error?

**Comment by dbeatty10** [comment]
> Is this difference causing a performance impact in dbt?

This is a savvy question to understand the problem holistically @Brooke-white 🧠

I'm not sure the precise performance impact is currently, but I did some quick experiments within https://github.com/dbt-labs/dbt-redshift/pull/468.

It feels like the two most important impacts on dbt users currently are:
1. That the error is even possible
1. When it does happen, the error message doesn't help the user understand what went wrong

We've tried to mitigate how commonly it might happen in https://github.com/dbt-labs/dbt-redshift/pull/468, but it's still possible for dbt users to create a seed file that would raise this error.

And it's definitely possible for non-dbt users to raise this same error. I don't know about other folks, but it took me a fair of research to understand what the error message actually meant (and to be honest, I still don't _totally_ understand 😅).

So raising a more understandable error message would be useful.

And figuring out how to make it less likely or impossible*, would be even better.

**Comment by Brooke-white** [comment]
Thanks for sharing those details @dbeatty10 , 😄 . 

Yes, I agree the error should not be possible & the error message should be easier to understand. But considering that I want to be cautious in implementing this method given that it is used in sqlalchemy, as tlocke mentioned in [the linked issue](https://github.com/tlocke/pg8000/issues/120#issuecomment-1293946586) by ensuring we have good test coverage in redshift-connector and sqlalchemy-redshift before merging. 

I believe the following is a path to fixing 1 (feel free to add anything else you think of)
- attempt plug/play with the pg8000 implementation
- run redshift-connector's test suite and see what breaks
- write tests for `setinputsizes` (likely can reuse much of pg8000 tests) for every datatype supported by Redshift (this means the datatypes we support via Postgres and the Redshift specific ones like `GEOMETRY`, `GEOGRAPHY`, `SUPER`, `VARBYTE`. FWIW we have a list of supported datatypes in type_utils file.

For fixing 2 I have a follow up question. As I haven't repro-d this myself, knowing exactly where this is thrown from redshift-connector could help us understand the options. If this is thrown by the driver, it is an easier case to catch it and repackage it with a more readable message than if it's given to us by Redshift server.

**Comment by soksamnanglim** [comment]
Hi @dataders and @dbeatty10,  
Thanks to everyone for providing notes from your investigation into this issue—it was very helpful 😄

Here is my investigation so far, 
Setinputsizes won't resolve the too many parameters limitation. To test: grabbed the prepared statement and parameters from the reprex using`insert_data_bulk` and the debugging stack _right_ before it generated the error. 
Executed the prepared statement against Redshift using pg8000, psycopg2, psycopg3, and redshift-connector. 
Ex:
`pg8000.exceptions.DatabaseError: {'S': 'ERROR', 'C': '42P02', 'M': 'there is no parameter $-32768', 'F': '../src/pg/src/backend/parser/parse_coerce.c', 'L': '279', 'R': 'coerce_type'}`

All drivers generated a similar error _except_ for psycopg2. I attempted to use setinputsizes with a very simple example: Insert 32768 rows of ints to a single column table. 32768 generates the same error with `setinputsizes`. 32767 executes successfully.

I attribute the root cause to the difference in protocols between psycopg2 and redshift-connector (+other drivers). Redshift-connector exclusively uses extended query protocol while psycopg2 only uses simple query. Pg8000 and Psycopg3 supports both. Psycopg2 merges parameters and prepared statement before it sends it to the database, see [here](https://github.com/psycopg/psycopg2/blob/329f43c76233ebb85f762f05f0b936f3ac5c223f/psycopg/cursor_type.c#L377C1-L459C2). Data is sent as string literals in the query which is why it never encountered the parameter limitation.

The parameter limitation 32767 is enforced on the client and server-side:
1. On the server side (for postgres), there is a hardcoded parameter limit for prepared statements. See it [here](https://github.com/postgres/postgres/blob/81ccbe520f42a9873734c614e4eb521190d8f8e4/src/interfaces/libpq/fe-exec.c#L1516C1-L1516C1). 
2. On the client side, the error is thrown when the driver attempts to convert the number of parameters to a signed 16-bit integer, causing overflow. So the database never receives the query. See [redshift-connector](https://github.com/aws/amazon-redshift-python-driver/blob/910be71314229e5b2febc60ddf8b8bcc992ea5f0/redshift_connector/core.py#L1672), and similarly, [pg8000](https://github.com/tlocke/pg8000/blob/f751b94a910975fd2e551814b7395336a5153cc8/pg8000/core.py#L639).

It likely is not just a driver issue as I changed the conversion to an unsigned int16 (pack format=`H`), hoping that it would allow double as many parameters, however I got the same error for pg8000 to Redshift. I believe the extended protocol for PARSE and BIND limits the number of parameters. 

As a baseline and quick fix:
1. I second @dbeatty10 @Brooke-white to make the error output more meaningful. Perhaps, classify it as [DataError](https://peps.python.org/pep-0249/#id31:~:text=DataError,of%20DatabaseError.). 

The fixes that require more time for implementation and testing can be broken down into two categories: low ~3-5 weeks and high time commitment ~2+ months. I still have to do some performance testing and will post my updates later. @jiezhen-chen

**Comment by dbeatty10** [comment]
Woah! Your investigation yielded amazing details @soksamnanglim 🤩

**Comment by Brooke-white** [comment]
fixed in 2.0.914

**Comment by dbeatty10** [comment]
🏆 @Brooke-white 

Is sounds like the new error message is something like this?

```
Prepared statement exceeds bind parameter limit 32767. 987654321 bind parameters were provided. Please retry with fewer bind parameters.
```

**Comment by Brooke-white** [comment]
Hey @dbeatty10, yep that's correct! @soksamnanglim addressed this in #187 and we are now raising a `DataError` with that message.
